proposal/WarClans612/proposal.rst

```diff
@@ -0,0 +1,171 @@
+
+Perspective users
+=================
+
+- Content Provider to generate tag for each content for easier recommendation
+- Social Media to generate tag to know the topic for each post (possibly prevent malicious usage or even for related content recommendation)
+- Forum Admin to generate tag for each content in its responsibility
+- Medical Worker to generate tag (e.g. possible diseases) from patient written record
+
+System architecture
+===================
+
+Input (when initializing):
+
+- Data containing past input and generated tag
+- All possible list of generated tag
+- If the model will be run on mobile device, we could improvise to shrink the model
+
+Output (when initialization):
+
+- Model that is ready to be used for generating tag
+
+Input (for usage after initialization):
+
+- Text content that needed tag generation
+
+Output (for usage after initialization):
+
+- List of possible tags
+
+Modularization
+-----------------
+
+- User API (only this layer is supposed to be accessed by user)
+- Normal model definition Module
+- Mobile enabled model definition Module
+- Training Module
+- Validation Module
+- Model saving module
+- Data Management Module
+- Config Management Module
+- Unittest Module
+- Example Module
+
+API description
+===============
+
+User will interact with the system using Python only.
+Using user provded data, the system will automatically take the provided data as
+training and let the model *learn*. After finishing the initialization, the model
+is then ready to be used.
+
+Possible API function:
+
+- generate_tag(string) -> return list_of_tag
+- use_mobile(True) -> enable smaller model for low powered device
+- possible_tag(list_of_string) -> user providing model with the possible output
+- training_data(pair_of_input_and_expected_output) -> user providing model with past knowledge
+
+Engineering infrastructure
+==========================
+
+By implementing the model from the paper, we would first define the model using PyTorch.
+Then building the training iteration and testing iteration modeule.
+Data Management Module will be implemented using Torch DataLoader.
+Testing each training iteration and results after training using PyTest.
+Each module will be provided with README.md file to describe each module purposes.
+Each function will be decorated with multi-line comment consisting of expected input,
+expected output, and possible requirements needed before using the function.
+
+Each new feature will be developed in the new branch.
+Ground Breaking and Critical Bug will have its own branch and being commited directly
+into newer version.
+Master branch version <1.0 will be initial implementation with no guarantee of it
+working at all.
+
+Schedule
+========
+
+- 05/01 - 05/07 -> Implementing Model
+- 05/08 - 05/15 -> Implementing Data Management Module
+- 05/16 - 05/24 -> Implementing Training Module
+- 05/25 - 05/30 -> Implementing Testing Module
+- 06/01 - 06/07 -> Implementing Unittesting
```

Comment:
If you make unit-testing early, and accumulate your testing code along the course of development, you will enjoy more productivity and fewer risks.  It's OK to list unit-testing here, though.

Reply:
I am actually thinking of the same thing. However, I have not thought about how to describet it in the schedule. Should I just remove it from the schedule and let it be part of the progress whenever I am implementing a new module?

Reply:
It's hard to say which way is better at this moment.  I'd suggest to keep it in mind that this is not yet fully clear when you implement.